File: packages/js/src/platform/domains.ts
import type { PlatformClient } from './client'

export interface DomainInfo {
  domain: string
  /** 'pending' until DNS propagates and the platform verifies the CNAME/TXT record. */
  verified: boolean
  txtRecord?: string
  cnameTarget?: string
}

export class DomainsAPI {
  constructor(private readonly client: PlatformClient) {}

  /**
   * Show the current domain(s) for a service.
   * The platform-assigned `.onlizard.com` subdomain is always present.
   */
  list(serviceId: string): Promise<DomainInfo[]> {
    return this.client.get(`/api/apps/${serviceId}/domains`)
  }

  /**
   * Attach a custom domain to a service.
   * Returns the DNS records to add before calling `verify()`.
   *
   * @example
   * ```ts
   * const info = await lizard.domains.add(serviceId, 'api.example.com')
   * console.log('Add CNAME:', info.cnameTarget)
   * ```
   */
  add(serviceId: string, domain: string): Promise<DomainInfo> {
    return this.client.post(`/api/apps/${serviceId}/domains`, { domain })
  }

  /**
   * Verify that DNS records have propagated and activate the domain.
   * Call after adding the CNAME or TXT record returned by `add()`.
   */
  verify(serviceId: string, domain: string): Promise<DomainInfo> {
    return this.client.post(`/api/apps/${serviceId}/domains/verify`, { domain })
  }
}

File: packages/python/lizard/platform/domains.py
from __future__ import annotations
from dataclasses import dataclass
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .client import PlatformClient


@dataclass
class DomainInfo:
    domain: str
    verified: bool
    cname_target: str | None = None
    service_id: str | None = None

    @classmethod
    def _from_dict(cls, d: dict) -> "DomainInfo":
        return cls(
            domain=d.get("domain", ""),
            verified=bool(d.get("verified", False)),
            cname_target=d.get("cnameTarget"),
            service_id=d.get("appId") or d.get("serviceId"),
        )


class DomainsAPI:
    def __init__(self, client: "PlatformClient") -> None:
        self._client = client

    def list(self, service_id: str) -> list[DomainInfo]:
        """List all custom domains for a service."""
        result = self._client.get(f"/api/apps/{service_id}/domains")
        return [DomainInfo._from_dict(d) for d in (result if isinstance(result, list) else result.get("domains", []))]

    def add(self, service_id: str, domain: str) -> DomainInfo:
        """Add a custom domain to a service."""
        return DomainInfo._from_dict(self._client.post(f"/api/apps/{service_id}/domains", {"domain": domain}))

    def verify(self, service_id: str, domain: str) -> DomainInfo:
        """Check DNS verification status for a custom domain."""
        return DomainInfo._from_dict(
            self._client.post(f"/api/apps/{service_id}/domains/verify", {"domain": domain})
        )
